### PR TITLE
refactor(rpc): remove intermediate types from rpc start up process

### PR DESCRIPTION
**Pull request by smatthewenglish**
the intermediary types `RpcServer`, `WsHttpServer`, `WsHttpServers`, and `WsHttpServerKind` are too rigid to support the generalization of rpc middleware. if they remained we'd need to specify every possible combination of middleware/server explicitly 

this pr removes them, laying the groundwork for configurable rpc middleware, as @mattsse and i discussed [here](https://github.com/paradigmxyz/reth/pull/7976)

### CODE REVIEW

**Review by emhane — COMMENTED** [review]
Would have appreciated if the commit messages were more descriptive, so it could be easier to follow along with the incremental impl. note there is the git commit flag `--fixup` which takes as value a previous commit, can be handy sometimes.

File: crates/node/builder/src/rpc.rs [hunk 1]
```diff
@@ -285,19 +285,20 @@ where
...
-        }
-        if let Some(addr) = handle.ws_local_addr() {
-            info!(target: "reth::cli", url=%addr, "RPC WS server started");
-        }
-        handle
-    });
+    let mut server_config = config.rpc.rpc_server_config();
+
+    let cloned_modules = modules.clone();
+    let launch_rpc = server_config.start_ws_http(&cloned_modules).await?;
```
Comment (mattsse) [inline crates/node/builder/src/rpc.rs:291]: does this no longer start the ipc server?

File: crates/rpc/rpc-builder/src/lib.rs [hunk 2]
```diff
@@ -254,11 +250,11 @@ where
...
     let module_config = module_config.into();
-    let server_config = server_config.into();
-    RpcModuleBuilder::new(provider, pool, network, executor, events, evm_config)
-        .build(module_config)
-        .start_server(server_config)
-        .await
+    let mut server_config = server_config.into();
+    let modules = RpcModuleBuilder::new(provider, pool, network, executor, events, evm_config)
+        .build(module_config);
+    let handle: RpcServerHandle = server_config.start_ws_http(&modules).await?;
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:256]: this two step shouldn't be required, this should remain one single chain build + start

File: crates/rpc/rpc-builder/src/lib.rs [hunk 3]
```diff
@@ -1289,10 +1278,13 @@ impl RpcServerConfig {
     /// Builds the ws and http server(s).
     ///
     /// If both are on the same port, they are combined into one server.
-    async fn build_ws_http(
+    pub async fn start_ws_http(
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:1281]: this is now the new `start` function?

then we should rename this to start, and perhaps we should consume the value now?

File: crates/rpc/rpc-builder/src/lib.rs [hunk 4]
```diff
@@ -1409,37 +1410,20 @@ impl RpcServerConfig {
...
-
-        if let Some(builder) = self.ipc_server_config {
-            let metrics = modules.ipc.as_ref().map(RpcRequestMetrics::ipc).unwrap_or_default();
-            let ipc_path =
-                self.ipc_endpoint.unwrap_or_else(|| constants::DEFAULT_IPC_ENDPOINT.into());
-            let ipc = builder
-                .set_rpc_middleware(IpcRpcServiceBuilder::new().layer(metrics))
-                .build(ipc_path);
-            server.ipc = Some(ipc);
-        }
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:1439]: where did the ipc setup go?

**Review by mattsse — CHANGES_REQUESTED** [review]
I think this is on the right track.

needs some fine tuning and we need to ensure we start ipc as well

File: crates/rpc/rpc-builder/src/lib.rs (hunk 2 above)
Comment (smatthewenglish) [reply crates/rpc/rpc-builder/src/lib.rs]: is the updated code what you had in mind?

File: crates/rpc/rpc-builder/src/lib.rs (hunk 3 above)
Comment (smatthewenglish) [reply crates/rpc/rpc-builder/src/lib.rs]: what do you mean by consume the value now

File: crates/rpc/rpc-builder/src/lib.rs (hunk 4 above)
Comment (smatthewenglish) [reply crates/rpc/rpc-builder/src/lib.rs:1439]: in my recent changes i've integrated it into the start method, thank you for pointing this out

File: crates/node/builder/src/rpc.rs [hunk 5]
```diff
@@ -288,19 +288,19 @@ where
...
+
+    if let Some(path) = launch_rpc.ipc_endpoint() {
+        info!(target: "reth::cli", %path, "RPC IPC server started");
+    }
+    if let Some(addr) = launch_rpc.http_local_addr() {
+        info!(target: "reth::cli", url=%addr, "RPC HTTP server started");
+    }
+    if let Some(addr) = launch_rpc.ws_local_addr() {
+        info!(target: "reth::cli", url=%addr, "RPC WS server started");
+    }
```
Comment (mattsse) [inline crates/node/builder/src/rpc.rs:303]: we can keep doing .map_ok and

```rust
let (rpc, auth) = futures::future::try_join(launch_rpc, launch_auth).await?;
```

so we can start the servers concurrently

File: crates/rpc/rpc-builder/src/lib.rs [hunk 6]
```diff
@@ -1298,13 +1285,17 @@ impl RpcServerConfig {
...
-    /// Builds the ws and http server(s).
+    /// Builds and starts the ws and http server(s).
     ///
     /// If both are on the same port, they are combined into one server.
-    async fn build_ws_http(
+    pub async fn start(
         &mut self,
         modules: &TransportRpcModules,
-    ) -> Result<WsHttpServer, RpcError> {
+    ) -> Result<RpcServerHandle, RpcError> {
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:1294]: cool, this deletes the intermediary wshttpserver step which was just bloart

File: crates/rpc/rpc-builder/src/lib.rs [hunk 7]
```diff
@@ -1298,13 +1285,17 @@ impl RpcServerConfig {
         self.jwt_secret.map(|secret| AuthLayer::new(JwtAuthValidator::new(secret)))
     }
 
-    /// Builds the ws and http server(s).
+    /// Builds and starts the ws and http server(s).
     ///
     /// If both are on the same port, they are combined into one server.
-    async fn build_ws_http(
+    pub async fn start(
         &mut self,
```
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:1292]: I think this should now just consume the type
```suggestion
        self,
```

File: examples/rpc-db/src/main.rs [hunk 8]
```diff
@@ -77,9 +77,9 @@ async fn main() -> eyre::Result<()> {
     server.merge_configured(custom_rpc.into_rpc())?;
 
     // Start the server & keep it alive
-    let server_args =
+    let mut server_args =
```
Comment (mattsse) [inline examples/rpc-db/src/main.rs:80]: this can be simplified if we don do &mut self on start

File: crates/rpc/rpc-builder/tests/it/utils.rs [hunk 9]
```diff
@@ -97,15 +89,11 @@ pub async fn launch_http_ws_same_port(modules: impl Into<RpcModuleSelection>) ->
...
-                .with_http(Default::default())
-                .with_http_address(addr),
-        )
-        .await
-        .unwrap()
+    let mut config = RpcServerConfig::ws(Default::default())
+        .with_ws_address(addr)
+        .with_http(Default::default())
+        .with_http_address(addr);
+    config.start(&server).await.unwrap()
```
Comment (mattsse) [inline crates/rpc/rpc-builder/tests/it/utils.rs:96]: same

**Review by mattsse — CHANGES_REQUESTED** [review]
cool, this is a lot cleaner now, no more gigantic type aliases.

**Comment by smatthewenglish** [comment]
> cool, this is a lot cleaner now, no more gigantic type aliases.

it also clears the way for generalizable HttpMiddleware/RpcMiddleware

**Review by mattsse — APPROVED** [review]
nice, lgtm!

**Pull request merged**